### PR TITLE
Implement complaint management for tutors

### PR DESCRIPTION
**Pull request by rpadovani**
### Checklist
- [x] I run `yarn run webpack:build:main`: the project builds without errors.
- [x] I run `yarn lint`: the project builds without code style warnings.
- [x] ~I updated the documentation and models.~
- [x] I tested the changes and all related features on the test server https://artemistest.ase.in.tum.de.
- [x] ~I added (end-to-end) test cases for the new functionality.~

### Motivation and Context
Implement the feature for managing students' complaints as tutors

Basically, tutors can see complaints about other tutors, but not about themselves.  They can leave a reply (the only way to resolve a complaint) and since the complaint is in the submission page, they can also override the submission, if it is needed.

This is just a first implementation, but it is fully functional

#### Student's view
screenshot-localhost-9000-2019 04 12-19-09-58

In the submission page, students can see their complaint, the response to their complaint, and the tutor who has replied to their complaint (if they want to followup or have other questions)

#### Tutor's view

Tutor course dashboard without any complaint:

screenshot-localhost-9000-2019 04 12-18-57-38

Tutor course dashboard with a complaint to evaluate:

tutor_dashboard_complaint

Tutor course dashboard with a complaint that has been evaluated, tutors can still read it but they cannot act over it

screenshot-localhost-9000-2019 04 12-19-23-20

Text assessment page where tutors can replay to a complaint:

screenshot-localhost-9000-2019 04 12-19-09-58

Text assessment page with a complaint already reviewed

screenshot-localhost-9000-2019 04 12-19-16-02

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/web/rest/ComplaintResponseResource.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+     * GET /complaint-responses/:id : get the "id" complaint response.
+     *
+     * @param id the id of the complaint response to retrieve
+     * @return the ResponseEntity with status 200 (OK) and with body the complaint response, or with status 404 (Not Found)
+     */
+    @GetMapping("/complaint-responses/{id}")
+    @PreAuthorize("hasAnyRole('USER', 'TA', 'INSTRUCTOR', 'ADMIN')")
+    public ResponseEntity<ComplaintResponse> getComplaintResponse(@PathVariable Long id) {
+        log.debug("REST request to get ComplaintResponse : {}", id);
+        Optional<ComplaintResponse> complaintResponse = complaintResponseRepository.findById(id);
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ComplaintResponseResource.java]: we need a check that the user is allowed to see this

File: src/main/java/de/tum/in/www1/artemis/web/rest/ComplaintResponseResource.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    public ResponseEntity<ComplaintResponse> createComplaint(@RequestBody ComplaintResponse complaintResponse) throws URISyntaxException {
+        log.debug("REST request to save ComplaintResponse: {}", complaintResponse);
+        if (complaintResponse.getId() != null) {
+            throw new BadRequestAlertException("A new complaint response cannot already have an id", ENTITY_NAME, "idexists");
+        }
+
+        if (complaintResponse.getComplaint() == null || complaintResponse.getComplaint().getId() == null) {
+            throw new BadRequestAlertException("A complaint response can be only associated to a complaint", ENTITY_NAME, "noresultid");
+        }
+
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ComplaintResponseResource.java]: we need a check that the tutor is allowed to create the complaint response

File: src/main/java/de/tum/in/www1/artemis/web/rest/ComplaintResponseResource.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    /**
+     * POST /complaint-responses: create a new complaint response
+     *
+     * @param complaintResponse the complaint response to create
+     * @return the ResponseEntity with status 201 (Created) and with body the new complaint response
+     * @throws URISyntaxException if the Location URI syntax is incorrect
+     */
+    @PostMapping("/complaint-responses")
+    @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
+    public ResponseEntity<ComplaintResponse> createComplaint(@RequestBody ComplaintResponse complaintResponse) throws URISyntaxException {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ComplaintResponseResource.java]: please rename to createComplaintResponse

File: src/main/java/de/tum/in/www1/artemis/web/rest/ComplaintResponseResource.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+    /**
+     * Get /complaint-responses/complaint/:id get a complaint response associated with the complaint "id"
+     *
+     * @param complaintId the id of the complaint for which we want to find a linked response
+     * @return the ResponseEntity with status 200 (OK) and with body the complaint response, or with status 404 (Not Found)
+     */
+    @GetMapping("/complaint-responses/complaint/{complaintId}")
+    @PreAuthorize("hasAnyRole('USER', 'TA', 'INSTRUCTOR', 'ADMIN')")
+    public ResponseEntity<ComplaintResponse> getComplaintResponseByResultId(@PathVariable Long complaintId) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ComplaintResponseResource.java]: The method name is confusing? How is the resultId used?

File: src/main/java/de/tum/in/www1/artemis/web/rest/ComplaintResponseResource.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        log.debug("REST request to get ComplaintResponse associated to complaint : {}", complaintId);
+        Optional<ComplaintResponse> complaintResponse = complaintResponseRepository.findByComplaint_Id(complaintId);
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ComplaintResponseResource.java]: we need a check if the user is allowed to see the complaint response.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ComplaintResponseResource.java (hunk 1 above)
Comment (rpadovani) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ComplaintResponseResource.java]: Done

File: src/main/java/de/tum/in/www1/artemis/web/rest/ComplaintResponseResource.java (hunk 2 above)
Comment (rpadovani) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ComplaintResponseResource.java]: Done

File: src/main/java/de/tum/in/www1/artemis/web/rest/ComplaintResponseResource.java (hunk 3 above)
Comment (rpadovani) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ComplaintResponseResource.java]: Right, thanks for noticing!

File: src/main/java/de/tum/in/www1/artemis/web/rest/ComplaintResponseResource.java (hunk 4 above)
Comment (rpadovani) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ComplaintResponseResource.java]: Yeah, misstype, is `getComplaintResponseByComplaintId`, thanks for noticing :)

File: src/main/java/de/tum/in/www1/artemis/web/rest/ComplaintResponseResource.java (hunk 5 above)
Comment (rpadovani) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ComplaintResponseResource.java]: Done

**Pull request merged**